### ISSUE TITLE
[2020] TabTransformer: Tabular Data Modeling Using Contextual Embeddings

### ISSUE THREAD

**Issue by cfiken** [cfiken/paper-reading#123]
## 0. 論文

| タイトル | TabTransformer: Tabular Data Modeling Using Contextual Embeddings |
|:-------|:--|
| リンク  | https://arxiv.org/abs/2012.06678 |
| 著者    | Xin Huang, Ashish Khetan, Milan Cvitkovic, Zohar Karnin |
| 投稿日付 | 2020/12/11 on arXiv |

## 1. どんなもの？

テーブルデータ向けの Transformer ベースモデルである TabTransformer を提案。入力特徴を離散値と連続値に分けて、離散値は embedding layers を通して Transformer に入力し、contextual embeddings を得、その後連続値特徴と concat して MLP レイヤに通し、ターゲットを予測する(下図)。
![スクリーンショット 2021-03-21 19 38 09](https://user-images.githubusercontent.com/1255485/111901742-fd965e00-8a7c-11eb-8545-451104db9ffd.png)

実験にて、(1) attention の可視化によりカテゴリ変数が(若干?)解釈可能であること、(2) noise data や missing data にロバストであること、(3) 既存のNNベースのモデルよりスコアが高く、GBDTと同等レベルのスコアであることを確認している。
(ただし、ベンチマークのスコアだけ見ると高いほうが多いのは GBDT モデル)

加えて、unlabeled data が多い場合の pretraining 手法も提案。BERT などと同様の MLM (Masked Language Model) 形式と、 ELECTRA と同様の RTD (Replaced Token Detection) 形式で実験し、unlabeled data がある程度あれば他の手法 (Pseudo Labeling, Entropy Regularization) と比べて精度を改善している。

## 2. 先行研究と比べてどこがすごい？

- 既存のNNベースの tabular データモデルのスコアを上回り、かつノイズや欠損に対してもロバスト
- unlabeled data がある場合の pretraining/finetuning の2ステージの学習方法も提案

## 3. 技術や手法のキモはどこ？

- 基本はシンプルな Transformer ベースのモデル
    - カテゴリの embeddings は、カテゴリ毎にユニークなベクトル c とカテゴリの各クラスごとのベクトル w を concat する形式で作成 `e = [c, w]`
        - concat ではなく add ( `e = c + w`; positional encoding と同じ) も試しており、Table 5 に結果があるが、concat の方が平均的に良いとして採用されている
        - その場合、embedding dim `d` に対してどの割合を c に当てるかどうかがハイパラになる (論文では 1/4 ~ 1/8 とか)
    - 学習した各カテゴリの embeddings を t-SNE で可視化し、ある程度似た特徴が近くに配置されているのを確認している
        - 下図の左が TabTransformer の最終層, 中央が TabTransformer に入れる前, 右がただの MLP に入れた後 (細かい情報は論文参照)
        - ![スクリーンショット 2021-03-21 19 56 34](https://user-images.githubusercontent.com/1255485/111902222-b9588d00-8a7f-11eb-9984-d8b38cf194eb.png)
- 半教師あり学習として TabTransformer-MLM と TabTransformer-RTD を提案
    - 前者はBERTと同様の Masked Language Model を各カテゴリ変数に使用したもの、後者は ELECTRA と同様の Replaced Token Detection を各カテゴリ変数に適用したもの
    - unlabeled データが多い場合にもデータの表現を学習できる

## 4. どうやって有効だと検証した？

15のベンチマークセットで実験しており、NNベースのモデルでは TabTransformer が安定してよく、GBDT と比べても同等に近い性能となっている。一番スコアが良いのは GBDT が多いが、全てのデータセットの平均AUCではほとんど変わらない。
下表が全ての平均、その下の表が全てのデータセットごとの結果の一部。
![スクリーンショット 2021-03-21 20 03 28](https://user-images.githubusercontent.com/1255485/111902391-9e3a4d00-8a80-11eb-9881-454866672a66.png)
![スクリーンショット 2021-03-21 20 02 16](https://user-images.githubusercontent.com/1255485/111902321-603d2900-8a80-11eb-8fed-b53dd6c7557a.png)

他のNNモデルに比べてノイズや欠損に対してロバストであることも確認している(下図の左がノイズ、右が欠損があるときの結果)。
![スクリーンショット 2021-03-21 20 05 21](https://user-images.githubusercontent.com/1255485/111902440-cf1a8200-8a80-11eb-873d-2c1d70ef2a71.png)

半教師あり学習の提案手法 (MLM, RTD) と他の手法 (Entropy Regularization, Pseudo Labeling, DAE) も比較しており、unlabeled データも finetuning に使える ER, PL と比べても良い結果になっている。
labeled data が少ない場合は ER, PL に比べて精度がやや落ちているが、良い表現が得られても分類層の学習に使えるデータが少ないためだと主張しており、MLM や RTD に更に ER や PL を組み合わせることで改善できるかもしれないとのこと。
表は上下の違いはデータ数が30k以上か未満かの違い。
![スクリーンショット 2021-03-21 20 09 32](https://user-images.githubusercontent.com/1255485/111902570-739cc400-8a81-11eb-960b-e8bf3377f8d4.png)

## 5. 議論はある？

ノイズなどにロバストであることから、contextual embeddings の中身の解釈をもっとちゃんとできそう。
image/text との組み合わせも色々考えられる。
ロバスト性については、GBDTモデルと比べてどうなのかは気になるところ(GBDTモデルはノイズや欠損に強いイメージ)。
